components/kyma-environment-broker/internal/orchestration/kyma/manager.go

```diff
@@ -200,37 +208,84 @@ func (u *upgradeKymaManager) updateOrchestration(o *internal.Orchestration, stat
 	return 0
 }
 
-func (u *upgradeKymaManager) waitForCompletion(o *internal.Orchestration) error {
-	// todo: use inter al config
-	// todo: remove PollInfinite  and introduce some timeout???
+// waitForCompletion waits until processing of given orchestration ends or if it's canceled
+func (u *upgradeKymaManager) waitForCompletion(o *internal.Orchestration, strategy orchestration.Strategy, execID string, log logrus.FieldLogger) (*internal.Orchestration, error) {
+	canceled := false
+	var err error
 	var stats map[domain.LastOperationState]int
-	err := wait.PollInfinite(u.pollingInterval, func() (bool, error) {
+	err = wait.PollInfinite(u.pollingInterval, func() (bool, error) {
+		// check if orchestration wasn't canceled
+		o, err = u.orchestrationStorage.GetByID(o.OrchestrationID)
+		switch {
+		case err == nil:
+			if o.State == orchestration.Canceled {
+				log.Info("Orchestration was canceled")
+				canceled = true
+			}
+		case dberr.IsNotFound(err):
+			log.Errorf("while getting orchestration: %v", err)
+			return false, err
+		default:
+			log.Errorf("while getting orchestration: %v", err)
+			return false, nil
+		}
 		s, err := u.operationStorage.GetOperationStatsForOrchestration(o.OrchestrationID)
 		if err != nil {
-			u.log.Errorf("while getting operations: %v", err)
+			log.Errorf("while getting operations: %v", err)
 			return false, nil
 		}
 		stats = s
 
-		numberOfInProgress, found := stats[domain.InProgress]
-		if !found {
-			u.log.Warnf("Orchestration %s operation stats does not contain in progress operations", o.OrchestrationID)
-			return true, nil
+		numberOfNotFinished := 0
+		numberOfInProgress, found := stats[orchestration.InProgress]
+		if found {
+			numberOfNotFinished += numberOfInProgress
+		}
+		numberOfPending, found := stats[orchestration.Pending]
+		if found {
+			numberOfNotFinished += numberOfPending
 		}
 
-		return numberOfInProgress == 0, nil
+		// don't wait for pending operations if orchestration was cancelled
+		if canceled {
+			return numberOfInProgress == 0, nil
+		} else {
+			return numberOfNotFinished == 0, nil
+		}
 	})
 	if err != nil {
-		return errors.Wrap(err, "while waiting for scheduled operations to finish")
+		return nil, errors.Wrap(err, "while waiting for scheduled operations to finish")
 	}
 
-	orchestrationState := orchestration.Succeeded
-	if stats[domain.Failed] > 0 {
-		orchestrationState = orchestration.Failed
+	if o.State == orchestration.Canceled {
+		strategy.Cancel(execID)
+		err = u.resolveCanceledOperations(o)
```

Comment:
I think the correct order is to first update cancelled state for each operation in DB, then signal `strategy.Cancel()`